### DISCUSSION TITLE
Using NOAA tile server in <UrlTile/> doesn't work - Zoom level doesn't map to TileMatrix correctly

### DISCUSSION THREAD

**Discussion by richcrad** [react-native-maps/react-native-maps#4354] [Q&A]
Hello!
I'm a big fan of React Native Maps, and have been using it in an app I'm developing to aid in sailing navigation.

I'm having an issue with Zoom level on a particular tile server (NOAA's Chart Services).

I want to use the [NOAA Chart Tools](https://gis.charttools.noaa.gov/arcgis/rest/services/MarineChart_Services/NOAACharts/MapServer) tile server to get a nautical charts overlay. You can use [NOAA's Online Map Viewer](https://noaa.maps.arcgis.com/home/webmap/viewer.html?url=https%3A%2F%2Fgis.charttools.noaa.gov%2Farcgis%2Frest%2Fservices%2FMarineChart_Services%2FNOAACharts%2FMapServer&source=sd) to see the charts.

Here's the URL template for the NOAA charts (from NOAA's documentation):
```
https://gis.charttools.noaa.gov/arcgis/rest/services/MarineChart_Services/NOAACharts/MapServer/WMTS/tile/1.0.0/MarineChart_Services_NOAACharts/{Style}/{TileMatrixSet}/{TileMatrix}/{TileRow}/{TileCol}.png
```
where:
```
{Style} = default
{TileMatrixSet} = GoogleMapsCompatible
and TileMatrix, TileRow, TileCol identify the tile in question
```

In my code, I use a UrlTile element with this urlTemplate:
```
const tileURL =
'https://gis.charttools.noaa.gov/arcgis/rest/services/MarineChart_Services/NOAACharts/MapServer/WMTS/tile/1.0.0/MarineChart_Services_NOAACharts/default/GoogleMapsCompatible/{z}/{y}/{x}.png';
```

Here's the problem: The Zoom level that gets passed to the tile server doesn't work. The zoom level that gets passed seems to be two levels higher than what the tile server is looking for.

To see what I mean, you can compare these two URLs:
[NOAA](https://gis.charttools.noaa.gov/arcgis/rest/services/MarineChart_Services/NOAACharts/MapServer/WMTS/tile/1.0.0/MarineChart_Services_NOAACharts/default/GoogleMapsCompatible/13/12654/5237.png) (uses zoom of 13)
[OpenStreetMap](https://tile.openstreetmap.org/15/5237/12654.png) (uses zoom of 15)

Note that they refer to the exact same tile, EXCEPT for the Zoom/TileMatrix. For fun, I hard-coded the {z} parameter to 13 in the urlTemplate in my app, then ran the app and zoomed-in until the tiles showed up correctly. So, I know the app is working except for the Zoom/TileMatrix parameter.

So, my question:
Is there any way to get the MapView/UrlTile to pass along the correct TileMatrix? Is there a way to "transform" it (e.g., subtract 2) before it calls the tile server?

Any help would be greatly appreciated! Thanks,
-Dick Craddock

**Comment by michaeladamkatz** [comment]
I am experiencing this also. Probably react people aren't the right group to ask about this NOAA-specific bug. I will try reporting it to NOAA.

**Reply by dockonit** [reply]
Have you had any reply from NOAA? I am having same issue

**Reply by michaeladamkatz** [reply]
I am in touch with the guy working on the tile set. He confirmed the off-by-2 problem. He sent me a beta to look at but it wasn't ready for prime time.

**Comment by dockonit** [comment]
FYI - I found this link in a GAIA forum - I have used for now to serve NOAA ENCs :
[https://gaiavectortilerendering.global.ssl.fastly.net/noaa-enc/{z}/{x}/{y}.png]

**Reply by michaeladamkatz** [reply]
Thanks. Do you know who publishes/pays for that?

**Reply by dockonit** [reply]
Sorry, no I don't. I have a free subscription level at Gaia, but am just using it for testing app during development ... hoping your guy comes thru at NOAA so they fix their link!! thx, s

**Comment by dockonit** [comment]
My Solution ... leave NOAA server's zoom as is, and just apply the offset in your code (as OP suggested).
In Swift:
            var NOAApath = path
            NOAApath.z = path.z - 2  // to offset NOAA ENC MapTile server by 2
            let request = URLRequest(url: self.url(forTilePath: NOAApath)) 

I've also discovered that I had to change the NOAA ENC maptile coordinate order from what OpenStreetMaps uses (z,x,y.png), to what works for NOAA (z,y,x.png): 

https://gis.charttools.noaa.gov/arcgis/rest/services/MarineChart_Services/NOAACharts/MapServer/WMTS/tile/1.0.0/MarineChart_Services_NOAACharts/default/default028mm/{z}/{y}/{x}.png

**Reply by electricpocket** [reply]
Thank you. This was driving me insane!!! Why doesn't NOAA update the description/xml
